src/anyio/abc/_sockets.py

```diff
@@ -31,6 +32,71 @@
 T_Retval = TypeVar("T_Retval")
 
 
+def _validate_socket(
+    sock_or_fd: socket.socket | int,
+    sock_type: socket.SocketKind,
+    addr_family: socket.AddressFamily = socket.AF_UNSPEC,
+    *,
+    require_connected: bool = False,
+    require_bound: bool = False,
+) -> socket.socket:
+    if isinstance(sock_or_fd, int):
+        try:
+            sock = socket.socket(fileno=sock_or_fd)
+        except OSError as exc:
+            if require_connected:
+                raise ValueError("the socket must be connected") from exc
+            elif require_bound:
+                raise ValueError("the socket must be bound to a local address") from exc
+            else:
+                raise
```

Comment:
This seems likely to cause confusion in cases where e.g. `exc.errno == errno.ENOTSOCK`, if the user pased the fd for a file rather than a socket.  Maybe use `exc.add_note()` rather than `raise ... from exc`?

Reply:
Good catch. I've added an explicit message (and an accompanying test) to cover this case.